### DISCUSSION TITLE
HTTP 419 session expired response after regenerating session

### DISCUSSION THREAD

**Discussion by zepfietje** [livewire/livewire#4332] [1. Help] [answered]
I have a reusable login form in a standalone Livewire component `AuthenticationForm` which emits a Livewire event `authenticated` after the user has logged in. This event is listened to by the page the authentication form is embedded on to execute further actions like redirecting to another page or saving the data the user entered in the flow on the page before logging in.

This all works fine, except one thing. The Laravel docs advise to regenerate the session after a successful login:
> [If authentication is successful, you should regenerate the user's session to prevent session fixation](https://laravel.com/docs/8.x/authentication#authenticating-users)

Unfortunately, this leads to a 419 HTTP response as the CSRF token gets updated due to the regenerated session. So the user gets a confirm dialog (https://github.com/livewire/livewire/pull/1146) and has to reload the page. This makes above-described flow impossible, since the emitted event never reaches the component listening for it.

Is there any proper way to handle the 419 HTTP response and achieve the flow I described?

**Comment by joshhanley** [comment]
@zepfietje hmm I'm not sure. Personally I probably wouldn't be using Livewire for my login form. Is there any reason you need that?

One thing you could do, is in the parent component, that listens for the `authenticated` event and does a redirect, it could regenerate the session and then do the redirect, so that way you don't have the CSRF issue then.

Hope this helps!

**Comment by zepfietje** [comment, marked as answer]
@joshhanley thanks for your thoughts!

I really do need Livewire for the login/register form, since it's quite a dynamic flow.

Yeah I'd been thinking about moving the session regeneration to the component listening for the `authenticated` event, but this would be bad practice security-wise.

What I ended up with is the following.

In `AuthenticationForm` Livewire component:

```php
session()->regenerate();
$this->emit('sessionRegenerated', csrf_token());
$this->emit('authenticated');
```

In `app.js`:

```js
Livewire.on('sessionRegenerated', (csrfToken) => {
  document.head
    .querySelector('meta[name="csrf-token"]')
    .setAttribute('content', csrfToken);
});
```

This way everything works just how I want it without having to redirect or running into CSRF issues.

**Reply by joshhanley** [reply]
@zepfietje no worries! 🙂 nice!
